### DISCUSSION TITLE
App context for model reflection causes circular import.

### DISCUSSION THREAD

**Discussion by TRezendes** [pallets-eco/flask-sqlalchemy#1138] [Q&A] [answered]
Although I was able to solve [my `RuntimeError: Working outside of application context.` issue](https://github.com/pallets-eco/flask-sqlalchemy/discussions/1133), that introduced a new one.

Now that I have to work inside an application context to reflect my database tables, I can't import my models into any of my blueprints without getting `ImportError: cannot import name 'blueprint' from partially initialized module 'app.blueprint' (most likely due to a circular import) (…APPNAME/appname_app/blueprint/__init__.py)`

models.py
```
from sqlalchemy.ext.hybrid import hybrid_property
from appname_app import argon2, db, create_app
from sqlalchemy import PrimaryKeyConstraint
from sqlalchemy.dialects import postgresql
from flask_login import UserMixin

app = create_app()
with app.app_context():
            
    class UserInfoTbl(db.Model):
        __tablename__ = 'user_info'
        __table_args__ = {
            'autoload':True,
            'autoload_with': db.engine
    }

etc.
```

/blueprint/__init__.py
```
from flask import Blueprint
from appname_app import login_manager
from flask import abort, Blueprint, flash, redirect, render_template, request, session, url_for
from appname_app.models import UserInfoTbl
from .blueprint_forms import LoginForm, RegistrationForm
from wtforms.validators import ValidationError
from flask_login import current_user, login_required, login_user, logout_user
from appname_app import argon2, db, login_manager
#from . import blueprint

@login_manager.user_loader
def user_loader(user_id):
    return UserInfoTbl.query.get(user_id)

@blueprint.route('/', methods=['GET', 'POST'])    
@blueprint.route('/login', methods=['GET', 'POST'])
def login():    
    if current_user.is_authenticated:
        flash(f'You are already logged in as {current_user.display_name}.', category='green')
        return redirect(url_for('players.index'))
        
    form = LoginForm()
    if form.validate_on_submit():
        attempted_user = db.session.execute(db.select(UserInfoTbl).filter_by(email=form.email.data)).first()
        if attempted_user and attempted_user.check_password(password=form.password.data):
            login_user(attempted_user)
            flash(f'Success! You are logged in as {attempted_user.display_name}.', category='green')
            return redirect(url_for('players.index'))
            
        else:
            flash('Username and password do not match! Please try again.', category='red')
            
    return render_template(
        '/blueprint/login.html',
        form = form
 
etc.
```

__init__.py
```
from flask import Flask, render_template, url_for, redirect, request
from flask_sqlalchemy import SQLAlchemy
from flask_wtf.csrf import CSRFProtect
from flask_login import LoginManager
from flask_argon2 import Argon2
import json

db = SQLAlchemy()
argon2 = Argon2()
csrf = CSRFProtect()
login_manager = LoginManager()


def create_app():
    app = Flask(__name__)
    
    with open('config.json') as config_file:
        config = json.load(config_file)

    # General configuration
    app.config['SECRET_KEY'] = config.get('SECRET_KEY')
    app.config['SQLALCHEMY_DATABASE_URI'] = config.get('DEV_DATABASE_URI')
    app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = config.get('SQLALCHEMY_TRACK_MODIFICATIONS')  

    db.init_app(app)
    db.app = app
   
    argon2.init_app(app)
    argon2.app = app

    csrf.init_app(app)
    csrf.app = app

    login_manager.init_app(app)
    login_manager.app = app

    login_manager.session_protection = 'strong'
    login_manager.login_view = 'security.login'
    login_manager.login_message_category = 'info'


    # Register blueprints
    from .blueprint import blueprint as blueprint

    app.register_blueprint(security, url_prefix='/blueprint')

    etc.
```

In short:

1. because I need an application context to reflect database tables, I have to import `create_app`

3. My blueprints are registered inside the definition of `create_app`.

5. In order to actually _use_ my models, I import them into my blueprints. 

7. And now we're back to 1.…

I've been banging my head against the wall days now. I feel like I must be missing something glaringly obvious. Any help would be much appreciated.

**Comment by davidism** [comment, marked as answer]
You need an active context when the code that needs it is evaluated. Module code is evaluated when it is imported. Therefore, you need a context active when you import the models, which happens as a side effect when you import the blueprints.

```python
def create_app():
    ...

    with app.app_context():
        from .blueprint import bp
        app.register_blueprint(bp)

    ...
    return app
```

**Reply by TRezendes** [reply]
Thank you so much. This appears to have solved not only the problem above, but also another problem I was having regarding blueprint registration.
